java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -261,6 +264,50 @@ void JNI_OnUnload(JavaVM* vm, void* reserved) {
   default_memory_pool_id = -1L;
 }
 
+/// Unpack the named tables passed through JNI.
+///
+/// Named tables are encoded as a string array, where every two elements 
+/// encode (1) the table name and (2) the address of an ArrowArrayStream
+/// containing the table data.  This function will eagerly read all
+/// tables into Tables.
+std::unordered_map<std::string, std::shared_ptr<arrow::Table>> LoadNamedTables(JNIEnv* env, jobjectArray& str_array) {
+  std::unordered_map<std::string, std::shared_ptr<arrow::Table>> map_table_to_record_batch_reader;
+  int length = env->GetArrayLength(str_array);
+  if (length % 2 != 0) {
+    JniThrow("Can not map odd number of array elements to key/value pairs");
+  }
+  std::shared_ptr<arrow::Table> output_table;
+  for (int pos = 0; pos < length; pos++) {
+    auto j_string_key = reinterpret_cast<jstring>(env->GetObjectArrayElement(str_array, pos));
+    pos++;
+    auto j_string_value = reinterpret_cast<jstring>(env->GetObjectArrayElement(str_array, pos));
+    long memory_address = 0;
+    try {
+      memory_address = std::stol(JStringToCString(env, j_string_value));
+    } catch (...) {
+      JniThrow("Failed to parse memory address from string value");
+    }
+    auto* arrow_stream_in = reinterpret_cast<ArrowArrayStream*>(memory_address);
+    std::shared_ptr<arrow::RecordBatchReader> readerIn = JniGetOrThrow(arrow::ImportRecordBatchReader(arrow_stream_in));
+    output_table = JniGetOrThrow(readerIn->ToTable());
+    map_table_to_record_batch_reader[JStringToCString(env, j_string_key)] = output_table;
+  }
+  return map_table_to_record_batch_reader;
+}
+
+/// Find the arrow Table associated with a given table name
+std::shared_ptr<arrow::Table> GetTableByName(const std::vector<std::string>& names,
+    std::unordered_map<std::string, std::shared_ptr<arrow::Table>> map_table_to_reader) {
+  std::shared_ptr<arrow::Table> output_table;
+  for (const auto& name : names) {
+    output_table = map_table_to_reader[name];
+    if (output_table == nullptr) {
+      JniThrow("Table name " + name + " is needed to execute the Substrait plan");
+    }
+  }
+  return output_table;
+}
```

Comment:
This still isn't correct. See earlier comments. I believe the test code in the repo is also incorrect.

Reply:
Oh no, tis is the main core of the proposal to map table name to table (on the same way that c++ test their join code). If you mention this isn't correct, then all proposal will be deleted.

Please I need you support to help me what part isn't correct? I'm trying to recover last messages and didn't see all the comments (I don't know why, maybe, I delete some on them).

- Unused input parameter?: will thrown a exception
- Plan references the same table twice?: recover the reference by name again
- Is there **some changes pending** o you recommend me to implement?
- Related to the test code, what I did to validate if Jar is working is install locally Dataset and run query agains that, like this for example:
```
import com.google.common.collect.ImmutableList;
import io.substrait.isthmus.SqlToSubstrait;
import io.substrait.proto.Plan;
import org.apache.arrow.dataset.file.FileFormat;
import org.apache.arrow.dataset.file.FileSystemDatasetFactory;
import org.apache.arrow.dataset.jni.NativeMemoryPool;
import org.apache.arrow.dataset.scanner.ScanOptions;
import org.apache.arrow.dataset.scanner.Scanner;
import org.apache.arrow.dataset.source.Dataset;
import org.apache.arrow.dataset.source.DatasetFactory;
import org.apache.arrow.dataset.substrait.AceroSubstraitConsumer;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.apache.calcite.sql.parser.SqlParseException;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

public class ClientSubstrait {
    public static void main(String[] args) {
        twoTables();
    }

    static Plan getPlanOnetable() throws SqlParseException {
        String sql = "SELECT * from nation";
        String nation = "CREATE TABLE NATION (N_NATIONKEY BIGINT NOT NULL, N_NAME CHAR(25), " +
                "N_REGIONKEY BIGINT NOT NULL, N_COMMENT VARCHAR(152))";
        SqlToSubstrait sqlToSubstrait = new SqlToSubstrait();
        Plan plan = sqlToSubstrait.execute(sql, ImmutableList.of(nation));
        return plan;
    }

    static Plan getPlanTwoTables() throws SqlParseException {
        String sql = "SELECT n.n_nationkey, n.n_name, c.c_name, c.c_phone, c.c_address FROM nation n JOIN customer c " +
                "ON n.n_nationkey = c.c_nationkey WHERE n.n_nationkey = 21";
        String nation = "CREATE TABLE NATION (N_NATIONKEY BIGINT NOT NULL, N_NAME CHAR(25), " +
                "N_REGIONKEY BIGINT NOT NULL, N_COMMENT VARCHAR(152))";
        String customer = "CREATE TABLE CUSTOMER (C_CUSTKEY BIGINT NOT NULL, C_NAME VARCHAR(25), " +
                "C_ADDRESS VARCHAR(40), C_NATIONKEY BIGINT NOT NULL, C_PHONE CHAR(15), C_ACCTBAL DECIMAL, " +
                "C_MKTSEGMENT CHAR(10), C_COMMENT VARCHAR(117) )";
        SqlToSubstrait sqlToSubstrait = new SqlToSubstrait();
        Plan plan = sqlToSubstrait.execute(sql, ImmutableList.of(nation, customer));
        return plan;
    }

    static void oneTable() {
        String uri = "s3://voltrondata-labs-datasets/nyc-taxi-tiny/year=2022/month=2/part-0.parquet";
        ScanOptions options = new ScanOptions(/*batchSize*/ 32768);
        try (
                BufferAllocator allocator = new RootAllocator();
                DatasetFactory datasetFactory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(),
                        FileFormat.PARQUET, uri);
                Dataset dataset = datasetFactory.finish();
                Scanner scanner = dataset.newScan(options);
                ArrowReader reader = scanner.scanBatches()
        ) {
            // map table to reader
            Map<String, ArrowReader> mapTableToArrowReader = new HashMap<>();
            mapTableToArrowReader.put("NATION", reader);
            // get binary plan
            Plan plan = getPlanOnetable();
            ByteBuffer substraitPlan = ByteBuffer.allocateDirect(plan.toByteArray().length);
            substraitPlan.put(plan.toByteArray());
            // run query
            try (ArrowReader arrowReader = new AceroSubstraitConsumer(allocator).runQuery(
                    substraitPlan,
                    mapTableToArrowReader
            )) {
                while (arrowReader.loadNextBatch()) {
                    System.out.println(arrowReader.getVectorSchemaRoot().contentToTSVString());
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    static void twoTables(){
        String uri1 = "file:///Users/dsusanibar/voltron/fork/consumer-testing/tests/data/tpch_parquet/nation.parquet";
        String uri2 = "file:///Users/dsusanibar/voltron/fork/consumer-testing/tests/data/tpch_parquet/customer.parquet";
        ScanOptions optionsNations = new ScanOptions(/*batchSize*/ 32768);
        ScanOptions optionsCustomer = new ScanOptions(/*batchSize*/ 32768);
        try (
                BufferAllocator allocator = new RootAllocator();
                DatasetFactory datasetFactory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(),
                        FileFormat.PARQUET, uri1);
                Dataset dataset = datasetFactory.finish();
                Scanner scanner = dataset.newScan(optionsNations);
                ArrowReader readerNation = scanner.scanBatches();
                DatasetFactory datasetFactoryCustomer = new FileSystemDatasetFactory(allocator,
                        NativeMemoryPool.getDefault(), FileFormat.PARQUET, uri2);
                Dataset datasetCustomer = datasetFactoryCustomer.finish();
                Scanner scannerCustomer = datasetCustomer.newScan(optionsCustomer);
                ArrowReader readerCustomer = scannerCustomer.scanBatches()
        ) {
            // map table to reader
            Map<String, ArrowReader> mapTableToArrowReader = new HashMap<>();
            mapTableToArrowReader.put("NATION", readerNation);
            mapTableToArrowReader.put("CUSTOMER", readerCustomer);
            // get binary plan
            Plan plan = getPlanTwoTables();
            ByteBuffer substraitPlan = ByteBuffer.allocateDirect(plan.toByteArray().length);
            substraitPlan.put(plan.toByteArray());
            // run query
            try (ArrowReader arrowReader = new AceroSubstraitConsumer(allocator).runQuery(
                    substraitPlan,
                    mapTableToArrowReader
            )) {
                while (arrowReader.loadNextBatch()) {
                    System.out.println(arrowReader.getVectorSchemaRoot().contentToTSVString());
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
```

Reply:
I've explained already: the implementation here makes no sense. Why are we looping through a list of names and just discarding the lookups? Both this code and the test case you linked to are not right. Substrait is trying to represent a hierarchical name like "catalog.schema.table". We don't support that, so as I already explained, the code should instead check that the names array is exactly of length 1, and only do a single lookup.

Reply:
Thank you for the detail. Let me add more detail how it is doing internally, let start the case when we need Join for two Tables:
```cpp
arrow::Status execute_substrait_low_level() {
    auto left_table = RunDatasetParquetScan("file:///Users/dsusanibar/voltron/fork/consumer-testing/substrait_consumer/data/tpch_parquet/nation.parquet");
    auto right_table = RunDatasetParquetScan2("file:///Users/dsusanibar/voltron/fork/consumer-testing/substrait_consumer/data/tpch_parquet/customer.parquet");
    std::map<std::string, std::shared_ptr<Table>> mapTableToMemoryAddress;
    mapTableToMemoryAddress["NATION"] = left_table;
    mapTableToMemoryAddress["CUSTOMER"] = right_table;
    // this table_provider was executed two times (alligned to the number of Tables into the Plan)
    arrow::engine::NamedTableProvider table_provider = [&mapTableToMemoryAddress](const std::vector<std::string>& names) {
        std::shared_ptr<arrow::Table> output_table;
        std::cout << "Tables names size : " << names.size() << std::endl; // alwais is "1" and it is handled by Acero C++ behavior
        for (const auto& name : names) { // this is a for but always iterarte once time
            output_table = mapTableToMemoryAddress[name];
        }
        // another way without for will be: output_table = mapTableToMemoryAddress[names.front()];
        std::shared_ptr<compute::ExecNodeOptions> options =
                std::make_shared<compute::TableSourceNodeOptions>(std::move(output_table));
        return compute::Declaration("table_source", {}, options, "mock_source");
    };
    arrow::engine::ConversionOptions conversion_options;
    conversion_options.named_table_provider = std::move(table_provider);
    ARROW_ASSIGN_OR_RAISE(std::string substrait_json, GetSubstraitJSONV21());
    ARROW_ASSIGN_OR_RAISE(auto buffer, arrow::engine::SerializeJsonPlan(substrait_json));
    Result<std::shared_ptr<RecordBatchReader>> reader = arrow::engine::ExecuteSerializedPlan(*buffer, NULLPTR, NULLPTR, conversion_options);
    auto table= arrow::Table::FromRecordBatchReader(reader.ValueOrDie().get()).ValueOrDie();
    std::cout << "Results : " << table->ToString() << std::endl;
}
```
Consider:
1. For 1 query that contais two tables:
- This is an incorrect behavior think that: `arrow::engine::NamedTableProvider table_provider = [&mapTableToMemoryAddress](const std::vector<std::string>& names)` is only executed once time, and, all the tables (nation/customer) are iterated into `for (const auto& name : names) {` two times.
- Correct behavior is: `arrow::engine::NamedTableProvider table_provider = [&mapTableToMemoryAddress](const std::vector<std::string>& names)` is executed any times of number of table are in Plan and `for (const auto& name : names) {` is all the time only iterated over one table all the time. This is the main reason because there is not a discarding of lookups.

2. The names array is exactly of length 1 it is handled/validates/passed by Acero.

3. Do a single lookup: We use a for but in real it is only once time, maybe we could change to `output_table = mapTableToMemoryAddress[names.front()];`

Please @lidavidm  let me know if with this explanation current implementation is clear or need some kind of additional changes.

I know there will be improvements, but as a Java JNI pass thru we are only consuming current defined API behavior.

Reply:
Sorry, but none of this is relevant.

Reply:
The function as implemented is not correct. Its job is to look up a single table. It should validate the input properly.

Reply:
```cpp
std::shared_ptr<arrow::Table> GetTableByName(const std::vector<std::string>& names,
    const std::unordered_map<std::string, std::shared_ptr<arrow::Table>>& tables) {
  if (names.size() != 1) {
    JniThrow("Tables with hierarchical names are not supported");
  }
  const auto& it = tables.find(names[0]);
  if (it == tables.end()) {
    JniThrow("Table is referenced, but not provided: " + names[0]);
  }
  return it->second;
}
```

Reply:
Your version is much easier to digest.